docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,99 @@
+# ADR ADR-001: Node Key Refactoring
+
+## Changelog
+
+- 2022-10-31: First draft
+
+## Status
+
+Proposed
+
+## Context
+
+The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.
+
+The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases:
+
+1. Rollback of the tree to a previous version
+2. Remove unnecessary old nodes
+
+## Decision
+
+- Use the version and the sequenced integer ID as a node key like `bigendian(version) | bigendian(nonce)` format. 
+- Remove the `version` field from node body writes.
+- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
+- Remove the `orphans` from the tree.
+
+New node structure
+
+```go
+type Node struct {
+	key           []byte
+	value         []byte
+	hash          []byte    // keep this field in the storage
+	nonce         int64     // new field, the sequenced integer ID within the specific version
+	leftNodeKey   int64     // new field, need to store in the storage
+	rightNodeKey  int64     // new field, need to store in the storage
+	version       int64     
+	size          int64
+	leftNode      *Node
+	rightNode     *Node
+	subtreeHeight int8
+	persisted     bool
+}
+```
+
+New tree structure
+
+```go
+type MutableTree struct {
+	*ImmutableTree                                    
+	lastSaved                *ImmutableTree
+	nonce                    int64                    // new field to track the integer ID of the current version
+	versions                 map[int64]bool           
+	allRootLoaded            bool                     
+	unsavedFastNodeAdditions map[string]*fastnode.Node
+	unsavedFastNodeRemovals  map[string]interface{}   
+	ndb                      *nodeDB
+	skipFastStorageUpgrade   bool 
+
+	mtx sync.Mutex
+}
+```
+
+### Migration
+
+We can migrate nodes one by one by iterating the version.
+
+- Iterate the version in order, and get the root node for the specific version.
+- Iterate the tree and assign the `nodeKey` to nodes which the node version equals. 
```

Comment:
Migrate a archive node and pruned node will end up with different node keys, but it's probably ok.

Reply:
Inconsistent node nonces could make the state-sync snapshot inconsistent as well?

Reply:
On the global nonce they would be different but version | local nonce should be same, right?

Reply:
Hmm, in migration, it depends on the order traversal, this is easy to specify. but when handling the insertion, how do we specify the logic, to allow alternative implementations be compatible with each other.

Reply:
I think the inconsistency of nodeKey is not a big problem, the only thing we need is to match the proof.

Reply:
We can test this on a live network with state sync and a simple migration to see if it works to be sure

Reply:
> I think the inconsistency of nodeKey is not a big problem, the only thing we need is to match the proof.

it depends on how we export the state sync snapshot I think, if we re-map the nodekey, then the stored one don't matter.

Reply:
> Hmm, in migration, it depends on the traversal order, this is easy to specify, like preorder+ascending. but when handling the insertion, we should also specify the precise logic of nonce assignments, to allow alternative implementations be compatible with each other.

I think the nonce assignment should happens in the `SaveBranch`, which do preorder ascending traversal, pruning historical branches.

Reply:
you are right